pkg/server/plugin/zmq/broker.go

```diff
@@ -221,50 +243,93 @@ func (lb *Broker) Channeler() {
 			lb.logger.Debugf("zmq/broker: zmq => channel %#x, %s\n", frames[0], frames)
 			// Dispacth back to the channel based on the zmq first frame
 			address := string(frames[0])
-			respChan, ok := lb.addressChan[address]
-			if !ok {
-				lb.logger.Infof("zmq/broker: chan not found for worker %s\n", address)
+			messageType := string(frames[2])
+			bounceCount, err := strconv.Atoi(string(frames[3]))
+			if err != nil {
+				lb.logger.Infof("zmq/broker: Cannot parse bounce_count int %v\n", frames)
 				break
 			}
-			delete(lb.addressChan, address)
-			respChan <- frames[2]
+			requestID := string(frames[4])
+			message := frames[6]
+			if messageType == Request {
+				parcel := newParcel(message)
+				parcel.Worker = address
+				lb.ReqChan <- parcel
+				go func (p *Parcel) {
+					response := <- p.RespChan
+					frames := [][]byte{
+						[]byte(address),
+						[]byte(address),
+						[]byte{},
+						[]byte(Response),
+						[]byte(strconv.Itoa(bounceCount)),
+						[]byte(requestID),
+						[]byte{},
+						response,
+					}
+					lb.respChan <- frames
+				}(parcel)
+			} else if messageType == Response {
+				respChan, ok := lb.addressChan[requestID]
+				if !ok {
+					lb.logger.Infof("zmq/broker: chan not found for worker %s\n", address)
+					break
+				}
+				delete(lb.addressChan, requestID)
+				respChan <- message
+			}
 		case p := <-lb.recvChan:
 			// Save the chan and dispatch the message to zmq
 			// If current no worker ready, will retry after HeartbeatInterval.
 			// Retry for HeartbeatLiveness times
-			address := lb.workers.Next()
+			// TODO(b123400): instead of asking for a new worker
+			// get the current worker (if possible, e.g. when nested, we already have the worker id)
+			// Or, we can put the worker back to pool when received request from plugin? not very good?
+			var address string
+			if p.Worker != "" {
+				address = p.Worker
+			} else {
+				address = lb.workers.Next()
+			}
 			if address == "" {
 				if p.retry < HeartbeatLiveness {
 					p.retry += 1
 					lb.logger.Infof("zmq/broker: no worker available, retry %d...\n", p.retry)
-					go func(p2 *parcel) {
+					go func(p2 *Parcel) {
 						time.Sleep(HeartbeatInterval)
 						lb.recvChan <- p2
 					}(p)
 					break
 				}
 				lb.logger.Infof("zmq/broker: no worker available, timeout.\n")
-				p.respChan <- []byte{0}
+				p.RespChan <- []byte{0}
 				break
 			}
 			addr := []byte(address)
+			requestID := generateRequestID()
 			frames := [][]byte{
 				addr,
 				addr,
 				[]byte{},
-				p.frame,
+				[]byte(Request),
+				[]byte("0"),
+				[]byte(requestID),
+				[]byte{},
+				p.Frame,
 			}
-			lb.addressChan[address] = p.respChan
+			lb.addressChan[requestID] = p.RespChan
```

Comment:
I would like to save the response channels by something like `"${request_id}-${bounce_count}"` which is more clear and we don't have to split id into parts.
(Bounce count is not implemented yet so you don't see it in code)

Splitting id into prefix-suffix means we will introduce another schema in the protocol, which is not necessary. The current design allows any form of request-id and worker-id, it can be a uuidv4 and everything will be fine, prefix-suffix just makes things easier to debug, but we should not enforce it in protocol level.